### PR TITLE
Prefix match in first iteration of beam search OP

### PR DESCRIPTION
**Pull request by viboga**
**Description**: This PR targets required changes to allow prefix matching in beam search OP.

**Motivation and Context**
- Why is this change required? What problem does it solve?
In generative models, prefix of the last word needs to be matched with the first word being generated.

- If it fixes an open issue, please link to the issue here.
NA

** Tests :**
Attached test report
[report.docx](https://github.com/microsoft/onnxruntime/files/7969498/report.docx)

### CODE REVIEW

File: onnxruntime/contrib_ops/cpu/transformers/dump_tensor.h [hunk 1]
```diff
@@ -26,7 +26,8 @@ namespace transformers {
     continue;                                                                               \
   }
 
-#define SKIP_IF_TOO_MANY(row_or_column_size, i, new_line) SKIP_IF_MORE_THAN(row_or_column_size, i, MAX_ROW_OR_COLUMN, new_line)
+//#define SKIP_IF_TOO_MANY(row_or_column_size, i, new_line) SKIP_IF_MORE_THAN(row_or_column_size, i, MAX_ROW_OR_COLUMN, new_line)
+#define SKIP_IF_TOO_MANY(row_or_column_size, i, new_line)
```
Comment (viboga) [inline onnxruntime/contrib_ops/cpu/transformers/dump_tensor.h:30]: undo before commit

**Comment by tianleiwu** [comment]
Why not add an attribute to indicate that vocab_mask is used in first step only (default is false)? In this way, the change could be small.

**Comment by viboga** [comment]
> Why not add an attribute to indicate that vocab_mask is used in first step only (default is false)? In this way, the change could be small.

@tianleiwu  Are you saying use the vocab_mask already present to prefix match instead of using it on every iteration?

**Comment by tianleiwu** [comment]
> > Why not add an attribute to indicate that vocab_mask is used in first step only (default is false)? In this way, the change could be small.
> 
> @tianleiwu Are you saying use the vocab_mask already present to prefix match instead of using it on every iteration?

Right. I think we only need add a new attribute (like is_prefix_vocab_mask with default value 0) so that user could choose whether using it on first iteration or every iteration.

**Comment by viboga** [comment]
> 

Wanted to make sure I understood your suggestion. The purpose of vocab_mask and prefix_mask is mutually exclusive.

The vocab_mask you have added is to limit the search space of the suggestions. This is a generic implementation which is an option at runtime. You can change the vocab_mask to change the search space/suggestions.

Prefix_mask on first iteration however is to match the last word. Indeed, it is very specific to an input string and tokenizer. Both can be used on a model as needed.

File: onnxruntime/contrib_ops/cpu/transformers/beam_search.cc [hunk 2]
```diff
@@ -292,6 +294,25 @@ Status BeamSearchImpl<T>::CheckInputs(const OpKernelContextInternal& context) {
     parameters_->vocab_mask = vocab_mask->DataAsSpan<int32_t>();
   }
 
+  const Tensor* prefix_vocab_mask = context.Input<Tensor>(9);
+  if (prefix_vocab_mask != nullptr) {
+    // prefix_vocab_mask is optional
+    const auto& vocab_mask_dims = prefix_vocab_mask->Shape().GetDims();
+    if (vocab_mask_dims.size() != 1) {
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cpu/transformers/beam_search.cc:301]: For prefix mask, you will need 2D: (batch_size, vocab_size).

File: onnxruntime/python/tools/transformers/convert_beam_search.py [hunk 3]
```diff
@@ -128,6 +128,17 @@ def parse_arguments(argv=None):
...
 
+    beam_search_group.add_argument('--vocab_size',
+                                    type=int,
+                                    required=False,
+                                    help="Vocab_size of the underlying model")
+
+    beam_search_group.add_argument('--prefix_vocab_mask',
+                                    required=False,
+                                    action='store_true',
+                                    help="This vocab mask applies only to first iteration, enable if last work in query might need auto complete")
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/convert_beam_search.py:139]: typo: work => word

File: onnxruntime/python/tools/transformers/convert_beam_search.py [hunk 4]
```diff
@@ -128,6 +128,17 @@ def parse_arguments(argv=None):
                                    default=1,
                                    help='Positive. >1 to penalize and <1 to encorage.')
 
+    beam_search_group.add_argument('--vocab_size',
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/convert_beam_search.py:131]: Do not need vocab_size input. It can be deduced from GPT-2 model.

File: onnxruntime/contrib_ops/cpu/transformers/logits_processor.h [hunk 5]
```diff
@@ -76,12 +76,24 @@ class VocabMaskLogitsProcessor : public ILogitsProcessor<T> {
   gsl::span<const int32_t> vocab_mask_;
 };
 
+template <typename T>
+class PrefixVocabMaskLogitsProcessor : public ILogitsProcessor<T> {
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cpu/transformers/logits_processor.h:80]: No need add a new class. I think you only need modify VocabMaskLogitsProcessor (like add a Boolean parameter in constructor etc.) since most logic is same.

File: onnxruntime/contrib_ops/cpu/transformers/logits_processor.cc [hunk 6]
```diff
@@ -9,6 +9,8 @@ namespace onnxruntime {
 namespace contrib {
 namespace transformers {
 
+static int beam_search_iteration;
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cpu/transformers/logits_processor.cc:12]: The iteration can be stored in ISequences (For example, let Sequences stored the start sequence length, then it can be deduced from current sequence length - start sequence length). In this way, there is no need to store it here.

**Comment by tianleiwu** [comment]
> > 
> 
> @tianleiwu Wanted to make sure I understood your suggestion. The purpose of vocab_mask and prefix_mask is mutually exclusive.
> 
> The vocab_mask you have added is to limit the search space of the suggestions. This is a generic implementation which is an option at runtime. You can change the vocab_mask to change the search space/suggestions.
> 
> Prefix_mask on first iteration however is to match the incomplete last word of the input. Indeed, it is very specific to an input string and tokenizer. Both can be used on a model as needed.

It depends on the usage. If there is need to have both prefix matching and bad word list. It is fine to separate them. If not, I think it is fine to consolidate them since prefix mask is just like vocab_mask applied only to first iteration.

File: onnxruntime/contrib_ops/cpu/transformers/logits_processor.cc (hunk 6 above)
Comment (viboga) [reply onnxruntime/contrib_ops/cpu/transformers/logits_processor.cc:12]: As discussed, added a comment here as it is more intuitive and readable to have a common var.

File: onnxruntime/contrib_ops/cpu/transformers/logits_processor.h (hunk 5 above)
Comment (viboga) [reply onnxruntime/contrib_ops/cpu/transformers/logits_processor.h:80]: Updated the current class.

File: onnxruntime/python/tools/transformers/convert_beam_search.py (hunk 4 above)
Comment (viboga) [reply onnxruntime/python/tools/transformers/convert_beam_search.py:131]: vocab size is used while creating masks. Model vocab size can vary based on model. For instance, deep suggest is trained with 50297 whereas GPT2 is 50257.

File: onnxruntime/contrib_ops/cpu/transformers/dump_tensor.h (hunk 1 above)
Comment (tianleiwu) [reply onnxruntime/contrib_ops/cpu/transformers/dump_tensor.h:30]: please undo it

File: onnxruntime/contrib_ops/cpu/transformers/logits_processor.cc [hunk 7]
```diff
@@ -146,6 +148,40 @@ void VocabMaskLogitsProcessor<T>::Process(const ISequences* /*sequences*/,
...
+  if (beam_search_iteration > 1) {
+    return;
+  }
+
+  // Process vocabulary mask and set tokens with mask value 0 to -inf.
+  T* p = next_token_scores.scores.data();
+
+  // next_token_scores shape (batch_size * num_beams, vocab_size)
+  int num_beams = next_token_scores.batch_beam_size / batch_size_;
+  assert(num_beams*batch_size_ == next_token_scores.batch_beam_size);
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cpu/transformers/logits_processor.cc:168]: format the code

File: onnxruntime/python/tools/transformers/convert_beam_search.py (hunk 4 above)
Comment (tianleiwu) [reply onnxruntime/python/tools/transformers/convert_beam_search.py:131]: It is possible to get vocab_size from shape of logits output of the onnx model.

File: onnxruntime/python/tools/transformers/convert_beam_search.py [hunk 8]
```diff
@@ -301,6 +323,11 @@ def convert_model(args):
 
 
 def test_model(args, use_vocab_mask: bool = False, sentences: List[str] = None):
+
+    if args.prefix_vocab_mask:
+        print("Skipping parity test as prefix vocab mask is not implemented by Huggin Face")
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/convert_beam_search.py:328]: typo: Huggingface

File: onnxruntime/contrib_ops/cpu/transformers/logits_processor.cc (hunk 6 above)
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cpu/transformers/logits_processor.cc:12]: It is better not be stored as a global variable. (Consider that two nodes executed at the same time, that means this code is not thread safe).

Could we pass the parameter in Process function instead?

**Review by tianleiwu — APPROVED** [review]

File: onnxruntime/contrib_ops/cpu/transformers/logits_processor.cc (hunk 6 above)
Comment (viboga) [reply onnxruntime/contrib_ops/cpu/transformers/logits_processor.cc:14]: https://msdata.visualstudio.com/Vienna/_sprints/backlog/ONNX%20Inference/Vienna/Nickel?workitem=1606691
